x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/output/AutodetectResultProcessor.java

```diff
@@ -310,6 +323,46 @@ void processResult(AutodetectResult result) {
         }
     }
 
+    void bulkPersistWithRetry(CheckedRunnable<JobResultsPersister.BulkIndexException> bulkRunnable) {
+        int attempts = 0;
+        while(attempts < maximumFailureRetries) {
+            try {
+                bulkRunnable.run();
+                return;
+            } catch (JobResultsPersister.BulkIndexException ex) {
+                if (isDeadOrDying()) {
+                    return;
+                }
+                final int currentAttempt = attempts;
+                LOGGER.trace(
+                    () -> new ParameterizedMessage("[{}] bulk persist failure on attempt [{}] ", jobId, currentAttempt),
+                    ex
+                );
+                attempts++;
+                try {
+                    double backOff = ((1 << attempts) - 1) / 2.0;
+                    Thread.sleep((int)(backOff * 100));
+                } catch (InterruptedException interrupt) {
+                    LOGGER.warn(
+                        () -> new ParameterizedMessage("[{}] failed bulk indexing of results after [{}] attempts", jobId, currentAttempt),
+                        ex
+                    );
+                    return;
+                }
+            } catch (Exception e) {
+                if (isDeadOrDying()) {
+                    throw e;
+                }
+                LOGGER.warn(new ParameterizedMessage("[{}] Error processing autodetect result", jobId), e);
+            }
+        }
+        bulkResultsPersister.clearBulkRequest();
```

Comment:
> Another thought is now with the bulk retries, the bulk request items do not get cleared after a failure.

I think we should also be removing successful items from the bulk request before retrying it, as that will also reduce the burden on the bulk threadpool.  That will have to be done in the `JobResultsPersister` class.

Reply:
> I think we should also be removing successful items from the bulk request before retrying it

This is tricky. I will look into trying to filter out the successes. It may be doable because all the indexing requests contain a provided doc ID.